File: common/src/main/java/net/impactdev/gts/common/messaging/messages/admin/ForceDeleteMessageImpl.java
package net.impactdev.gts.common.messaging.messages.admin;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import net.impactdev.gts.api.GTSService;
import net.impactdev.gts.api.listings.Listing;
import net.impactdev.gts.api.listings.auctions.Auction;
import net.impactdev.gts.api.listings.buyitnow.BuyItNow;
import net.impactdev.gts.api.communication.message.errors.ErrorCode;
import net.impactdev.gts.api.communication.message.errors.ErrorCodes;
import net.impactdev.gts.api.communication.message.type.admin.ForceDeleteMessage;
import net.impactdev.gts.common.messaging.GTSMessagingService;
import net.impactdev.gts.common.messaging.messages.AbstractMessage;
import net.impactdev.gts.common.plugin.GTSPlugin;
import net.impactdev.impactor.api.json.factory.JObject;
import net.impactdev.impactor.api.utilities.printing.PrettyPrinter;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public abstract class ForceDeleteMessageImpl extends AbstractMessage implements ForceDeleteMessage {

    protected final UUID listing;
    protected final UUID actor;
    protected final boolean give;

    public ForceDeleteMessageImpl(UUID id, UUID listing, UUID actor, boolean give) {
        super(id);
        this.listing = listing;
        this.actor = actor;
        this.give = give;
    }

    @Override
    public @NonNull UUID getListingID() {
        return this.listing;
    }

    @Override
    public @NonNull UUID getActor() {
        return this.actor;
    }

    @Override
    public boolean shouldGive() {
        return this.give;
    }

    public static class ForceDeleteRequest extends ForceDeleteMessageImpl implements Request {

        public static final String TYPE = "Admin/Delete/Request";

        public static ForceDeleteRequest decode(@Nullable JsonElement content, UUID id) {
            if(content == null) {
                throw new IllegalStateException("Raw JSON data was null");
            }

            JsonObject raw = content.getAsJsonObject();

            UUID listing = Optional.ofNullable(raw.get("listing"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate listing ID"));
            UUID actor = Optional.ofNullable(raw.get("actor"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate actor ID"));
            boolean give = Optional.ofNullable(raw.get("give"))
                    .map(JsonElement::getAsBoolean)
                    .orElseThrow(() -> new IllegalStateException("Failed to locate give parameter"));

            return new ForceDeleteRequest(id, listing, actor, give);
        }

        public ForceDeleteRequest(UUID id, UUID listing, UUID actor, boolean give) {
            super(id, listing, actor, give);
        }

        @Override
        public CompletableFuture<ForceDeleteMessage.Response> respond() {
            return GTSPlugin.instance().storage().processForcedDeletion(this);
        }

        @Override
        public @NonNull String asEncodedString() {
            return GTSMessagingService.encodeMessageAsString(
                    TYPE,
                    this.getID(),
                    new JObject()
                            .add("listing", this.listing.toString())
                            .add("actor", this.actor.toString())
                            .add("give", this.give)
                            .toJson()
            );
        }

        @Override
        public void print(PrettyPrinter printer) {
            printer.kv("Message ID", this.getID())
                    .kv("Listing ID", this.listing)
                    .kv("Actor", this.actor);
        }
    }

    public static class ForceDeleteResponse extends ForceDeleteMessageImpl implements Response {

        public static final String TYPE = "Admin/Delete/Response";

        public static ForceDeleteResponse decode(@Nullable JsonElement content, UUID id) {
            if(content == null) {
                throw new IllegalStateException("Raw JSON data was null");
            }

            JsonObject raw = content.getAsJsonObject();

            UUID request = Optional.ofNullable(raw.get("request"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate or parse request ID"));
            UUID listing = Optional.ofNullable(raw.get("listing"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate listing ID"));
            UUID actor = Optional.ofNullable(raw.get("actor"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate actor ID"));
            Listing data = Optional.ofNullable(raw.get("data"))
                    .map(x -> {
                        JsonObject json = (JsonObject) x;
                        String type = json.get("type").getAsString();
                        if(type.equals("bin")) {
                            return GTSService.getInstance().getGTSComponentManager()
                                    .getListingResourceManager(BuyItNow.class)
                                    .get()
                                    .getDeserializer()
                                    .deserialize(json);
                        } else {
                            return GTSService.getInstance().getGTSComponentManager()
                                    .getListingResourceManager(Auction.class)
                                    .get()
                                    .getDeserializer()
                                    .deserialize(json);
                        }
                    })
                    .orElseThrow(() -> new IllegalStateException("Unable to locate or parse listing data"));
            boolean give = Optional.ofNullable(raw.get("give"))
                    .map(JsonElement::getAsBoolean)
                    .orElseThrow(() -> new IllegalStateException("Failed to locate give parameter"));
            boolean successful = Optional.ofNullable(raw.get("successful"))
                    .map(JsonElement::getAsBoolean)
                    .orElseThrow(() -> new IllegalStateException("Failed to locate success parameter"));
            ErrorCode error = Optional.ofNullable(raw.get("error"))
                    .map(x -> ErrorCodes.get(x.getAsInt()))
                    .orElse(null);

            return new ForceDeleteResponse(id, request, listing, actor, data, give, successful, error);
        }

        private final UUID request;

        private final Listing data;

        private final boolean successful;
        private final ErrorCode error;

        private long responseTime;

        public ForceDeleteResponse(UUID id, UUID request, UUID listing, UUID actor, Listing data, boolean give, boolean successful, @Nullable ErrorCode error) {
            super(id, listing, actor, give);
            this.request = request;
            this.data = data;
            this.successful = successful;
            this.error = error;
        }

        @Override
        public UUID getRequestID() {
            return this.request;
        }

        @Override
        public long getResponseTime() {
            return this.responseTime;
        }

        @Override
        public void setResponseTime(long millis) {
            this.responseTime = millis;
        }

        @Override
        public boolean wasSuccessful() {
            return this.successful;
        }

        @Override
        public Optional<ErrorCode> getErrorCode() {
            return Optional.ofNullable(this.error);
        }

        @Override
        public @NonNull String asEncodedString() {
            return GTSMessagingService.encodeMessageAsString(
                    TYPE,
                    this.getID(),
                    new JObject()
                            .add("request", this.getRequestID().toString())
                            .add("listing", this.listing.toString())
                            .add("actor", this.actor.toString())
                            .add("data", this.data.serialize())
                            .add("give", this.give)
                            .add("successful", this.successful)
                            .consume(o -> this.getErrorCode().ifPresent(e -> o.add("error", e.ordinal())))
                            .toJson()
            );
        }

        @Override
        public void print(PrettyPrinter printer) {
            printer.kv("Message ID", this.getID())
                    .kv("Request ID", this.getRequestID())
                    .kv("Listing ID", this.listing)
                    .kv("Actor", this.actor)
                    .kv("Should Return", this.give);
        }

        @Override
        public Optional<Listing> getDeletedListing() {
            return Optional.of(this.data);
        }

        public static class ForcedDeleteResponseBuilder implements ForceDeleteMessage.Response.ResponseBuilder {

            private UUID request;
            private UUID listing;
            private UUID actor;
            private Listing data;
            private boolean give;
            private boolean successful;
            private ErrorCode error;

            @Override
            public ResponseBuilder request(UUID request) {
                this.request = request;
                return this;
            }

            @Override
            public ResponseBuilder listing(UUID listing) {
                this.listing = listing;
                return this;
            }

            @Override
            public ResponseBuilder actor(UUID actor) {
                this.actor = actor;
                return this;
            }

            @Override
            public ResponseBuilder data(Listing data) {
                this.data = data;
                return this;
            }

            @Override
            public ResponseBuilder give(boolean give) {
                this.give = give;
                return this;
            }

            @Override
            public ResponseBuilder successful(boolean successful) {
                this.successful = successful;
                return this;
            }

            @Override
            public ResponseBuilder error(ErrorCode error) {
                this.error = error;
                return this;
            }

            @Override
            public ForceDeleteMessage.Response build() {
                return new ForceDeleteResponse(
                        GTSPlugin.instance().messagingService().generatePingID(),
                        this.request,
                        this.listing,
                        this.actor,
                        this.data,
                        this.give,
                        this.successful,
                        this.error
                );
            }
        }
    }
}

File: common/src/main/kotlin/net/impactdev/gts/common/messaging/messages/admin/ForceDeleteMessageImpl.kt
package net.impactdev.gts.common.messaging.messages.admin

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import net.impactdev.gts.api.GTSService.Companion.instance
import net.impactdev.gts.api.listings.Listing
import net.impactdev.gts.api.listings.auctions.Auction
import net.impactdev.gts.api.listings.buyitnow.BuyItNow
import net.impactdev.gts.api.messaging.message.errors.ErrorCode
import net.impactdev.gts.api.messaging.message.errors.ErrorCodes
import net.impactdev.gts.api.messaging.message.type.admin.ForceDeleteMessage
import net.impactdev.gts.api.util.PrettyPrinter
import net.impactdev.gts.common.messaging.GTSMessagingService
import net.impactdev.gts.common.messaging.messages.AbstractMessage
import net.impactdev.gts.common.plugin.GTSPlugin
import net.impactdev.impactor.api.json.factory.JObject
import java.util.*
import java.util.concurrent.CompletableFuture

abstract class ForceDeleteMessageImpl(
    id: UUID,
    override val listingID: UUID,
    override val actor: UUID,
    protected val give: Boolean
) : AbstractMessage(id), ForceDeleteMessage {
    override fun shouldGive(): Boolean {
        return give
    }

    class ForceDeleteRequest(id: UUID, listing: UUID, actor: UUID, give: Boolean) :
        ForceDeleteMessageImpl(id, listing, actor, give), ForceDeleteMessage.Request {
        override fun respond(): CompletableFuture<ForceDeleteMessage.Response?>? {
            return GTSPlugin.Companion.getInstance().getStorage().processForcedDeletion(this)
        }

        override fun asEncodedString(): String {
            return GTSMessagingService.Companion.encodeMessageAsString(
                TYPE,
                id,
                JObject()
                    .add("listing", listingID.toString())
                    .add("actor", actor.toString())
                    .add("give", give)
                    .toJson()
            )
        }

        override fun print(printer: PrettyPrinter) {
            printer.kv("Message ID", id)
                .kv("Listing ID", listingID)
                .kv("Actor", actor)
        }

        companion object {
            const val TYPE = "Admin/Delete/Request"
            @kotlin.jvm.JvmStatic
            fun decode(content: JsonElement?, id: UUID): ForceDeleteRequest {
                checkNotNull(content) { "Raw JSON data was null" }
                val raw = content.asJsonObject
                val listing = Optional.ofNullable(raw["listing"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate listing ID") }
                val actor = Optional.ofNullable(raw["actor"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate actor ID") }
                val give = Optional.ofNullable(raw["give"])
                    .map { obj: JsonElement -> obj.asBoolean }
                    .orElseThrow { IllegalStateException("Failed to locate give parameter") }
                return ForceDeleteRequest(id, listing, actor, give)
            }
        }
    }

    class ForceDeleteResponse(
        id: UUID,
        override val requestID: UUID?,
        listing: UUID?,
        actor: UUID?,
        private val data: Listing?,
        give: Boolean,
        private val successful: Boolean,
        private val error: ErrorCode?
    ) : ForceDeleteMessageImpl(id, listing!!, actor!!, give), ForceDeleteMessage.Response {
        override var responseTime: Long = 0
        override fun wasSuccessful(): Boolean {
            return successful
        }

        override val errorCode: Optional<ErrorCode>
            get() = Optional.ofNullable(error)

        override fun asEncodedString(): String {
            return GTSMessagingService.Companion.encodeMessageAsString(
                TYPE,
                id,
                JObject()
                    .add("request", requestID.toString())
                    .add("listing", listingID.toString())
                    .add("actor", actor.toString())
                    .add("data", data!!.serialize())
                    .add("give", give)
                    .add("successful", successful)
                    .consume { o: JObject -> errorCode.ifPresent { e: ErrorCode -> o.add("error", e.ordinal()) } }
                    .toJson()
            )
        }

        override fun print(printer: PrettyPrinter) {
            printer.kv("Message ID", id)
                .kv("Request ID", requestID)
                .kv("Listing ID", listingID)
                .kv("Actor", actor)
                .kv("Should Return", give)
        }

        override val deletedListing: Optional<Listing>
            get() = Optional.of(data)

        class ForcedDeleteResponseBuilder : ForceDeleteMessage.Response.ResponseBuilder {
            private var request: UUID? = null
            private var listing: UUID? = null
            private var actor: UUID? = null
            private var data: Listing? = null
            private var give = false
            private var successful = false
            private var error: ErrorCode? = null
            override fun request(request: UUID?): ForceDeleteMessage.Response.ResponseBuilder? {
                this.request = request
                return this
            }

            override fun listing(listing: UUID?): ForceDeleteMessage.Response.ResponseBuilder? {
                this.listing = listing
                return this
            }

            override fun actor(actor: UUID?): ForceDeleteMessage.Response.ResponseBuilder? {
                this.actor = actor
                return this
            }

            override fun data(data: Listing?): ForceDeleteMessage.Response.ResponseBuilder? {
                this.data = data
                return this
            }

            override fun give(give: Boolean): ForceDeleteMessage.Response.ResponseBuilder? {
                this.give = give
                return this
            }

            override fun successful(successful: Boolean): ForceDeleteMessage.Response.ResponseBuilder? {
                this.successful = successful
                return this
            }

            override fun error(error: ErrorCode?): ForceDeleteMessage.Response.ResponseBuilder? {
                this.error = error
                return this
            }

            override fun from(response: ForceDeleteMessage.Response): ForceDeleteMessage.Response.ResponseBuilder? {
                return this
            }

            override fun build(): ForceDeleteMessage.Response? {
                return ForceDeleteResponse(
                    GTSPlugin.Companion.getInstance().getMessagingService().generatePingID(),
                    request,
                    listing,
                    actor,
                    data,
                    give,
                    successful,
                    error
                )
            }
        }

        companion object {
            const val TYPE = "Admin/Delete/Response"
            @kotlin.jvm.JvmStatic
            fun decode(content: JsonElement?, id: UUID): ForceDeleteResponse {
                checkNotNull(content) { "Raw JSON data was null" }
                val raw = content.asJsonObject
                val request = Optional.ofNullable(raw["request"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate or parse request ID") }
                val listing = Optional.ofNullable(raw["listing"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate listing ID") }
                val actor = Optional.ofNullable(raw["actor"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate actor ID") }
                val data = Optional.ofNullable(raw["data"])
                    .map { x: JsonElement ->
                        val json = x as JsonObject
                        val type = json["type"].asString
                        if (type == "bin") {
                            return@map instance!!.gTSComponentManager
                                .getListingResourceManager(BuyItNow::class.java)
                                .get()
                                .deserializer
                                .deserialize(json)
                        } else {
                            return@map instance!!.gTSComponentManager
                                .getListingResourceManager(Auction::class.java)
                                .get()
                                .deserializer
                                .deserialize(json)
                        }
                    }
                    .orElseThrow { IllegalStateException("Unable to locate or parse listing data") }
                val give = Optional.ofNullable(raw["give"])
                    .map { obj: JsonElement -> obj.asBoolean }
                    .orElseThrow { IllegalStateException("Failed to locate give parameter") }
                val successful = Optional.ofNullable(raw["successful"])
                    .map { obj: JsonElement -> obj.asBoolean }
                    .orElseThrow { IllegalStateException("Failed to locate success parameter") }
                val error = Optional.ofNullable(raw["error"])
                    .map { x: JsonElement -> ErrorCodes[x.asInt] }
                    .orElse(null)
                return ForceDeleteResponse(id, request, listing, actor, data, give, successful, error)
            }
        }
    }
}